### PR TITLE
LMS Rails 5.0 Prework - Refactor ActiveRecord::Base.connection.execute calls to use TypeMapByOID, Part 3

### PR DESCRIPTION
**Pull request by brendanshean**
## WHAT
This is part 3 of https://github.com/empirical-org/Empirical-Core/pull/7711

There is the final part of `ActiveRecord::Base.connection.execute` replacements.
Refactor `ActiveRecord::Base.connection.execute` calls in Rails 4 into a new call RawSqlRunner.execute that fills in the appropriate TypeMap so that the resulting PG:Result matches the one in Rails 5.

## WHY
ActiveRecord::Base.connection.execute yields a PG::Result. Calling to_a on that PG::Result gives different values in Rails 4 and Rails 5. This seems to be due to the TypeMap associated with the PG::Result which is used in casting values.

### Notion Card Links
https://www.notion.so/quill/Refactor-ActiveRecord-Base-connection-execute-calls-to-use-TypeMapByOID-5b1a320b11824b7088a63458667891e8

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  YES
Have you deployed to Staging? |  YES
Self-Review: Have you done an initial self-review of the code below on Github? |
Design Review: If applicable, have you compared the coded design to the mockups? | N/A

### CODE REVIEW

File: services/QuillLMS/app/controllers/profiles_controller.rb [hunk 1]
```diff
@@ -118,8 +118,8 @@ def staff
     # the order in which the the query returns these.
     can_display_next_activity = begin
       @act_sesh_records.any? &&
-      @act_sesh_records.first['locked'] == 'f' &&
-      @act_sesh_records.first['marked_complete'] == 'f' &&
+      @act_sesh_records.first['locked'] == false &&
+      @act_sesh_records.first['marked_complete'] == false &&
```
Comment (emilia-friedberg) [inline services/QuillLMS/app/controllers/profiles_controller.rb:122]: YAY

File: services/QuillLMS/app/controllers/teachers/units_controller.rb [hunk 2]
```diff
@@ -218,61 +225,103 @@ def score_info
...
+            ua.created_at,
+            unit_activity_id,
+            state.completed,
+            ua.id
+            #{completed}
+          ORDER BY
+            ua.order_number,
+            unit_activity_id
+        SQL
+      ).to_a
```
Comment (emilia-friedberg) [inline services/QuillLMS/app/controllers/teachers/units_controller.rb]: holy hell

File: services/QuillLMS/client/app/bundles/Teacher/components/progress_reports/student_overview.jsx [hunk 3]
```diff
@@ -116,7 +116,7 @@ export default class extends React.Component {
         newRow['Average Score'] = countAndAverage.average;
         csvReportData.push(newRow);
       });
-      downloadReportOrLoadingIndicator = <CSVDownloadForProgressReport data={csvReportData} preserveCasing={'t'} />;
+      downloadReportOrLoadingIndicator = <CSVDownloadForProgressReport data={csvReportData} preserveCasing={true} />;
```
Comment (emilia-friedberg) [inline services/QuillLMS/client/app/bundles/Teacher/components/progress_reports/student_overview.jsx:119]: We might want to double check that this works - I think it should but this isn't something that's coming out of the query.

**Review by emilia-friedberg — APPROVED** [review]
LGTM! One tiny thing I think we should check but otherwise all good. Thanks for doing this!

File: services/QuillLMS/app/controllers/teachers/units_controller.rb (hunk 2 above)
Comment (brendanshean) [reply services/QuillLMS/app/controllers/teachers/units_controller.rb]: Transitioning this one to heredoc allows for nice complexity hiding:

Screen Shot 2021-06-01 at 11 38 16 AM

**Review by eadams17 — APPROVED** [review]

File: services/QuillLMS/client/app/bundles/Teacher/components/progress_reports/student_overview.jsx (hunk 3 above)
Comment (brendanshean) [reply services/QuillLMS/client/app/bundles/Teacher/components/progress_reports/student_overview.jsx:119]: I did a spot check of two files generated one with 't' and one with `true`

File: services/QuillLMS/app/controllers/teachers/units_controller.rb [hunk 4]
```diff
@@ -218,61 +225,103 @@ def score_info
...
+        if unit['assigned_student_ids'] && classroom_student_ids
+          active_assigned_student_ids = JSON.parse(unit['assigned_student_ids']) & classroom_student_ids
+          unit['number_of_assigned_students'] = active_assigned_student_ids.length
+        else
+          unit['number_of_assigned_students'] = 0
         end
+        unit
+      end
     else
       []
```
Comment (dandrabik) [inline services/QuillLMS/app/controllers/teachers/units_controller.rb]: Not necessary, but a chance for an early return above, e.g. 
```ruby
return [] if teach_own_or_coteach_classrooms_array.empty?

File: services/QuillLMS/app/models/blog_post.rb [hunk 5]
```diff
@@ -135,7 +135,14 @@ def add_published_at
...
     # This looks for slugs that look like #{current-slug}-2 so we
     # can change our slug for this post to increment the end digit.
-    existing_posts_with_incremented_slug = ActiveRecord::Base.connection.execute("SELECT slug FROM blog_posts WHERE slug ~* CONCAT(#{ActiveRecord::Base.sanitize(slug)}, '-\\d$');").to_a.map{ |h| h['slug'] }
+    existing_posts_with_incremented_slug = RawSqlRunner.execute(
+      <<-SQL
+        SELECT slug
+        FROM blog_posts
+        WHERE slug ~* CONCAT(#{ActiveRecord::Base.sanitize(slug)}, '-\\d$')
+      SQL
+    ).values.flatten
```
Comment (dandrabik) [inline services/QuillLMS/app/models/blog_post.rb:144]: nice cleanup!

File: services/QuillLMS/app/queries/lesson_recommendations_query.rb [hunk 6]
```diff
@@ -27,25 +27,43 @@ def activity_recommendations
...
+      <<-SQL
+        SELECT DISTINCT unit.name
+        FROM units unit
+        LEFT JOIN classroom_units AS cu
+          ON cu.unit_id = unit.id
+        WHERE cu.classroom_id = #{classroom_id}
+          AND cu.visible = true
+          AND unit.visible = true
+      SQL
+    ).to_a.map {|e| e['name']}
```
Comment (dandrabik) [inline services/QuillLMS/app/queries/lesson_recommendations_query.rb]: Not that important, but you could do the `values.flatten` here as well for consistency.

**Review by dandrabik — APPROVED** [review]
LGTM! Thanks for cleaning this up. The SQL is much easier to read on multiple lines. Also, the javascript cleans up so nicely being about to use native types.

I left a few small comments that you can take or leave.

**Comment by brendanshean** [comment]
@eadams17 @emilia-friedberg I added a couple of commits:   [Update javascript test data](Update javascript test data) and [[Update snapshots](https://github.com/empirical-org/Empirical-Core/pull/7731/commits/abb3ddace12ce0071c3debf3ac41e8e98a36a93e).  The diff is quite large with those two commits, but it's basically replacing test data string ids with integer ones and then updating the corresponding jest snapshots.

**Review by emilia-friedberg — APPROVED** [review]
LGTM!

**Pull request merged**